# Pandas Trick for Daily Productivity

In [59]:
import seaborn as sns
import pandas as pd
import numpy as np

## 1. Create a Built-in Dataset

- Use `seaborn` library to generate a quick dataset for experimental purpose.

In [78]:
# See the available ready-to-use datasets here
print(sns.get_dataset_names())

['anagrams', 'anscombe', 'attention', 'brain_networks', 'car_crashes', 'diamonds', 'dots', 'exercise', 'flights', 'fmri', 'gammas', 'geyser', 'iris', 'mpg', 'penguins', 'planets', 'taxis', 'tips', 'titanic']


In [60]:
df = sns.load_dataset('mpg')
df.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0    3504          12.0   
1  15.0          8         350.0       165.0    3693          11.5   
2  18.0          8         318.0       150.0    3436          11.0   
3  16.0          8         304.0       150.0    3433          12.0   
4  17.0          8         302.0       140.0    3449          10.5   

   model_year origin                       name  
0          70    usa  chevrolet chevelle malibu  
1          70    usa          buick skylark 320  
2          70    usa         plymouth satellite  
3          70    usa              amc rebel sst  
4          70    usa                ford torino

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   origin        398 non-null    object 
 8   name          398 non-null    object 
dtypes: float64(4), int64(3), object(2)
memory usage: 28.1+ KB


In [62]:
df.describe()

mpg   cylinders  displacement  horsepower       weight  \
count  398.000000  398.000000    398.000000  392.000000   398.000000   
mean    23.514573    5.454774    193.425879  104.469388  2970.424623   
std      7.815984    1.701004    104.269838   38.491160   846.841774   
min      9.000000    3.000000     68.000000   46.000000  1613.000000   
25%     17.500000    4.000000    104.250000   75.000000  2223.750000   
50%     23.000000    4.000000    148.500000   93.500000  2803.500000   
75%     29.000000    8.000000    262.000000  126.000000  3608.000000   
max     46.600000    8.000000    455.000000  230.000000  5140.000000   

       acceleration  model_year  
count    398.000000  398.000000  
mean      15.568090   76.010050  
std        2.757689    3.697627  
min        8.000000   70.000000  
25%       13.825000   73.000000  
50%       15.500000   76.000000  
75%       17.175000   79.000000  
max       24.800000   82.000000

## 2. Alternative `LIKE` SQL Method in Pandas

In [63]:
df[df['name'].str.contains('pinto')].head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
32   25.0          4          98.0         NaN    2046          19.0   
61   21.0          4         122.0        86.0    2226          16.5   
80   22.0          4         122.0        86.0    2395          16.0   
112  19.0          4         122.0        85.0    2310          18.5   
130  26.0          4         122.0        80.0    2451          16.5   

     model_year origin                 name  
32           71    usa           ford pinto  
61           72    usa  ford pinto runabout  
80           72    usa      ford pinto (sw)  
112          73    usa           ford pinto  
130          74    usa           ford pinto

In [64]:
# similar to df[df['name'].str.contains('^amc')]
df[df['name'].str.startswith('amc')].head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
3   16.0          8         304.0       150.0    3433          12.0   
9   15.0          8         390.0       190.0    3850           8.5   
16  18.0          6         199.0        97.0    2774          15.5   
24  21.0          6         199.0        90.0    2648          15.0   
33  19.0          6         232.0       100.0    2634          13.0   

    model_year origin                name  
3           70    usa       amc rebel sst  
9           70    usa  amc ambassador dpl  
16          70    usa          amc hornet  
24          70    usa         amc gremlin  
33          71    usa         amc gremlin

In [65]:
# similar to df[df['name'].str.contains('pinto$')]
df[df['name'].str.endswith('pinto')].head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
32   25.0          4          98.0         NaN    2046          19.0   
112  19.0          4         122.0        85.0    2310          18.5   
130  26.0          4         122.0        80.0    2451          16.5   
168  23.0          4         140.0        83.0    2639          17.0   
174  18.0          6         171.0        97.0    2984          14.5   

     model_year origin        name  
32           71    usa  ford pinto  
112          73    usa  ford pinto  
130          74    usa  ford pinto  
168          75    usa  ford pinto  
174          75    usa  ford pinto

## 3. Alternative `IN` SQL Method in Pandas

In [66]:
eighties = [80, 81, 82]
df[df['model_year'].isin(eighties)]
# similar to df[df['model_year'] >= 80]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
309  41.5          4          98.0        76.0    2144          14.7   
310  38.1          4          89.0        60.0    1968          18.8   
311  32.1          4          98.0        70.0    2120          15.5   
312  37.2          4          86.0        65.0    2019          16.4   
313  28.0          4         151.0        90.0    2678          16.5   
..    ...        ...           ...         ...     ...           ...   
393  27.0          4         140.0        86.0    2790          15.6   
394  44.0          4          97.0        52.0    2130          24.6   
395  32.0          4         135.0        84.0    2295          11.6   
396  28.0          4         120.0        79.0    2625          18.6   
397  31.0          4         119.0        82.0    2720          19.4   

     model_year  origin                   name  
309          80  europe              vw rabbit  
310          80   japan  toyota corolla tercel  
311          80     usa     chevrolet chevette  
312          80   japan             datsun 310  
313          80     usa     chevrolet citation  
..          ...     ...                    ...  
393          82     usa        ford mustang gl  
394          82  europe              vw pickup  
395          82     usa          dodge rampage  
396          82     usa            ford ranger  
397          82     usa             chevy s-10  

[89 rows x 9 columns]

## 4. Replace Values in Column Based on Condition

### 4.a. Replace a Single Value with a Single Condition

- Let's inspect the the __mpg__ of __datsun 310__ japanese car. Notice the value of it's mpg is 37.2.

In [67]:
df.loc[df['name'] == 'datsun 310']

mpg  cylinders  displacement  horsepower  weight  acceleration  \
312  37.2          4          86.0        65.0    2019          16.4   

     model_year origin        name  
312          80  japan  datsun 310

- Now, we will try to modify the value using `.loc` function. See that the value is changed from __37.2__ to __38.2__.

In [48]:
df.loc[df['name'] == 'datsun 310', 'mpg'] = 38.2
df.loc[df['name'] == 'datsun 310']

mpg  cylinders  displacement  horsepower  weight  acceleration  \
312  38.2          4          86.0        65.0    2019          16.4   

     model_year origin        name  
312          80  japan  datsun 310

### 4.b. Replace Multiple Value with Multiple Condition

- Now, we will change the value of origin for __Chevrolet Chevette__ from __usa__ to __asian__ specifically only for such car with displacement 98.0.

In [68]:
df[(df['name'] == 'chevrolet chevette') & (df['displacement'] == 98.0)]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
237  30.5          4          98.0        63.0    2051          17.0   
266  30.0          4          98.0        68.0    2155          16.5   
311  32.1          4          98.0        70.0    2120          15.5   

     model_year origin                name  
237          77    usa  chevrolet chevette  
266          78    usa  chevrolet chevette  
311          80    usa  chevrolet chevette

In [70]:
df.loc[(df['name'] == 'chevrolet chevette') & (df['displacement'] == 98.0), 'origin'] = 'asian'
df[(df['name'] == 'chevrolet chevette') & (df['displacement'] == 98.0)]

mpg  cylinders  displacement  horsepower  weight  acceleration  \
237  30.5          4          98.0        63.0    2051          17.0   
266  30.0          4          98.0        68.0    2155          16.5   
311  32.1          4          98.0        70.0    2120          15.5   

     model_year origin                name  
237          77  asian  chevrolet chevette  
266          78  asian  chevrolet chevette  
311          80  asian  chevrolet chevette

In [71]:
df.origin.unique()

array(['usa', 'japan', 'europe', 'asian'], dtype=object)

### 4.c. Other Possible Methods for Value Replacement
Source: Ref: https://sparkbyexamples.com/pandas/pandas-replace-values-based-on-condition/

In [ ]:
# Below are complete examples to replace values of columns in pandas.
import pandas as pd
import numpy as np
technologies = {
    'Courses':["Spark","PySpark","Python","pandas"],
    'Fee' :[20000,25000,22000,30000],
    'Duration':['30days','40days','35days','50days'],
    'Discount':[1000,2300,1200,2000]
              }
index_labels=['r1','r2','r3','r4']
df = pd.DataFrame(technologies,index=index_labels)
print(df)

# Replace values of columns by using DataFrame.loc[] property.
df2=df.loc[df['Fee'] > 22000, 'Fee'] = 1
print(df2)

# Using DataFrame.astype() Replace values based on condition.
df['Fee'] = (df['Fee'] > 22000).astype(int)
print(df)

# Replace values of Given column by using np.where() function.
df['Fee'] = np.where(df['Fee'] > 22000, 1, df['Fee'])
print(df)

# Using DataFrame.mask() function.
df['Fee'].mask(df['Fee'] >=22000 ,'0', inplace=True)
print(df)

## 5. Iterate the Value of Dataframe by Row

In [81]:
dummy = []
for idx, row in df.iterrows(): 
  if idx <= 5:
    dummy.append((row['model_year'], row['origin'], row['name']))
  else:
    break

dummy

[(70, 'usa', 'chevrolet chevelle malibu'),
 (70, 'usa', 'buick skylark 320'),
 (70, 'usa', 'plymouth satellite'),
 (70, 'usa', 'amc rebel sst'),
 (70, 'usa', 'ford torino'),
 (70, 'usa', 'ford galaxie 500')]